pandas/core/arrays/base.py

```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
         used for interacting with our indexers.
         """
         return np.array(self)
+
+
+def ExtensionOpsMixin(include_arith_ops, include_logic_ops):
+    """A mixin factory for creating default arithmetic and logical operators,
+    which are based on the underlying dtype backing the ExtensionArray
+
+    Parameters
+    ----------
+    include_arith_ops : boolean indicating whether arithmetic ops should be
+                        created
+    include_logic_ops : boolean indicating whether logical ops should be
+                        created
+
+    Returns
+    -------
+    A mixin class that has the associated operators defined.
+
+    Usage
+    ------
+    If you have defined a subclass MyClass(ExtensionArray), then
+    use MyClass(ExtensionArray, ExtensionOpsMixin(True, True)) to
+    get both the arithmetic and logical operators
+    """
+    class _ExtensionOpsMixin:
+        pass
+
+    def create_method(op_name):
+        def _binop(self, other):
+            def convert_values(parm):
+                if isinstance(parm, ExtensionArray):
+                    ovalues = list(parm)
+                elif is_extension_array_dtype(parm):
+                    ovalues = parm.values
+                elif is_list_like(parm):
+                    ovalues = parm
+                else:  # Assume its an object
+                    ovalues = [parm] * len(self)
+                return ovalues
+            lvalues = convert_values(self)
+            rvalues = convert_values(other)
+
+            # Get the method for each object.
+            def callfunc(a, b):
+                f = getattr(a, op_name, None)
+                if f is not None:
+                    return f(b)
+                else:
+                    return NotImplemented
+            res = [callfunc(a, b) for (a, b) in zip(lvalues, rvalues)]
```

Comment:
If we would do `op(a, b)` directly (so in practice doing `a + b` instead of `a.__add__(b)`), you would not need the checking for NotImplemented below?

Reply:
similar as https://github.com/pandas-dev/pandas/blob/c85ab083919b59ce84c220d5baf7d34ff4a0bcf2/pandas/core/indexes/category.py#L793

Reply:
The check for `NotImplemented` is needed because the underlying `ExtensionDtype` class (which are the objects in `lvalues` and `rvalues` above) may not implement the specific operator.

Not sure what you are referring to when you wrote "similar as" the code in `pandas/core/indexes/category.py".  That implementation is not doing things element by element, which is what I'm doing here.

Reply:
why are you doing it this way, dynamically generating this? there is a limited set of ops, just write them out.

Reply:
> why are you doing it this way, dynamically generating this? there is a limited set of ops, just write them out.

@jreback are you speaking about this `create_method` function that you commented on? Or the fact below the `create_method` function is used in a loop instead of manually repeated? 
Because if it is the first: that is how we do it almost everywhere in the pandas codebase.

Reply:
https://github.com/pandas-dev/pandas/pull/21191/commits/21d76b322c6e5ba87f95a9ac21a17e55b1915cc2

is what I mean.

Reply:
@jorisvandenbossche My latest commit does as you suggest.  This removes the tests for `NotImplemented`

@jreback My latest commit uses a similar pattern to what you had in the referenced commit, just creating an operator one by one instead of using a loop